p4-16/psa/PSA.mdk

```diff
@@ -309,6 +310,134 @@ Additional supported match_kind types
+advantage of this difference, could save 10 Mbits of storage in a
+table with a million entries[^CacheCost].
+
+[^CacheCost]: While 10 Mbits sounds tiny to one accustomed to computers with hundreds of gigabytes of DRAM, the highest speed PSA implementations are ASICs that must keep tables in on-chip memories, similar to caches in general purpose CPUs.  The Intel i9-7980XE released in 2017 has 198 Mbits of on-chip L3 cache shared by its CPU cores.  Among Intel processors in Intel's 7th generation Core released in 2017 with at least 100 Mbits of L3 cache, they all cost close to $9 per Mbit of L3 cache.  https://en.wikipedia.org/wiki/List_of_Intel_microprocessors
+
+The P4 Runtime API is planning to use 64-bit quantities to hold values
+of the types listed in section [#sec-psa-type-definitions].  For
+control plane operations on tables, any trimming or padding of values
+will be performed in the agent (trimming from controller to device
+direction, padding in device to controller direction).
+
+There are multiple channels of communication over which such values
+might be carried between the controller and the data plane.  These
+channels of communication include:
+
++ Control plane operations on tables, where values of these types may
+  be included as part of the key, or as an action parameter.
++ Packets sent to the CPU ("packet in" from the controller's
+  perspective), or received from the CPU ("packet out" from the
+  controller's perspective).
++ Fields in a `Digest` extern notification message (section
+  [#sec-packet-digest]).
++ Fields in the data contents of a `Register` array (section
+  [#sec-registers]).
++ Match fields of an entry in a `ValueSet` extern instance (section
+  [#sec-parser-value-sets]).
+
+Note: There may be other channels not listed above.
+
+For packets between the control plane and the PSA device, there is the
+issue that many PSA implementations are expected to restrict P4
+programs to have headers with fields with a total length that is a
+multiple of 8 bits.  To make it possible to define P4 header types
+that meet this restriction, and contain values of fields with these
+PSA-specific types, and be source-compatible across multiple PSA
+implementations, additional types are defined that contain `InHeader` in
+their name.  For example, `PortIdInHeader_t` is the same as `PortId_t`,
+except it must be a multiple of 8 bits long, and contain at least as
+many bits as `PortId_t` does.
+
+Because these types are guaranteed to
+be a multiple of 8 bits long, any combination of them may be included
+in a P4 header type definition, as long as the other fields in the
+header satisfy the multiple of 8 bits restriction.  The controller or
+P4 program generating packets with such headers should fill in any
+most significant "padding" bits with 0.  In a P4 program, this can be
+done with a normal assignment statement, where the value on the right
+hand side is cast to the wider `InHeader` type.
```

Comment:
suggest: Similarly, for an assignment of a PortIdInHeader_t field to a PortId_t field, there will be a truncation of the excess most significant bits as part of the cast.

Reply:
I added a sentence like this.  Thanks.